{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "2_Biopython.ipynb",
      "provenance": [],
      "collapsed_sections": [],
      "authorship_tag": "ABX9TyOAqxXr0gsu+ZqW0aU7iCcd",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/Julibio7/bioinformatics/blob/main/2_Biopython.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "# <center> Iniciando o Biopython"
      ],
      "metadata": {
        "id": "j7ANo5xRsDV1"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "- [tutorial oficial Biopython](http://biopython.org/DIST/docs/tutorial/Tutorial.html#sec7)\n",
        "- [Cookbook](https://biopython.org/wiki/Category%3ACookbook)\n",
        "-[Biopython.org](https://biopython.org/)\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "q02-GJhBzVpX"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Primeiro √© necess√°rio instalar o Biopython no Colab usando o comando:"
      ],
      "metadata": {
        "id": "Lv9Wk9FLsgv_"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#Instalar o Biopython\n",
        "!pip3 install biopython"
      ],
      "metadata": {
        "id": "-AVaotzYsRXI"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#importar Bio.seq\n",
        "import Bio\n",
        "from Bio.Seq import Seq"
      ],
      "metadata": {
        "id": "wJl78zphrPYj"
      },
      "execution_count": 3,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "GI5fYHe6rJDW"
      },
      "outputs": [],
      "source": [
        "#Declarando minha primeira sequ√™ncia\n",
        "minha_sequencia = Seq(\"AGTACACTGGT\")\n",
        "print(minha_sequencia)"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "##Manipulando sequ√™ncias\n",
        "Uma das fun√ß√µes b√°sicas do Biopython √© a an√°lise de sequ√™ncias. Para isso, √© necess√°rio importar o m√≥dulo Seq com o comando:"
      ],
      "metadata": {
        "id": "ubk50aInsxMM"
      }
    },
    {
      "cell_type": "code",
      "source": [
        " # Manipula√ß√£o b√°sica de sequ√™ncias\n",
        "from Bio.Seq import Seq "
      ],
      "metadata": {
        "id": "JY_2O-Ias6wU"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "Em Bioinform√°tica, nucleot√≠deos presentes em sequ√™ncias de DNA s√£o representados pelos caracteres A, T, C e G (adenina, timina, citosina e guanina)."
      ],
      "metadata": {
        "id": "q0fEMUASvuHv"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Em Biopython, sequ√™ncias devem ser declaradas com o m√©todo `Seq( )`:"
      ],
      "metadata": {
        "id": "mnfEWME8tB6y"
      }
    },
    {
      "cell_type": "code",
      "source": [
        " # Declarando uma sequ√™ncia\n",
        "seq1 = Seq(\"ACGTAGCTACGATCACAGCTA\")\n",
        "print(\"Minha sequ√™ncia √©\", seq1)"
      ],
      "metadata": {
        "id": "uKQLRjiZtBLI",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "3906f74b-ac1e-4635-fc8f-613a31c9f7b4"
      },
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Minha sequ√™ncia √© ACGTAGCTACGATCACAGCTA\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        " Assim, define-se como sequ√™ncia complementar a sequ√™ncia formada por bases que pareiam com bases de outra fita de DNA, ou seja, bases que realizam liga√ß√µes de hidrog√™nio na forma√ß√£o da dupla h√©lice do DNA"
      ],
      "metadata": {
        "id": "5dvfVsk1vjz2"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Ap√≥s declarada, a sequ√™ncia pode ser manipulada como um objeto do tipo Seq. Observe por exemplo, como obter o reverso complementar da sequ√™ncia declarada anteriormente:\n",
        "\n"
      ],
      "metadata": {
        "id": "nbPWlHSDtPEQ"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "rc = seq1.reverse_complement()\n",
        "print(\"O reverso complementar √©\", rc)"
      ],
      "metadata": {
        "id": "lpFGQUzTtSZq"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Podemos ainda obter a sequ√™ncia transcrita chamando o m√©todo `transcribe()`:"
      ],
      "metadata": {
        "id": "tY46yjm6tXlT"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Transcri√ß√£o\n",
        "rna = seq1.transcribe()\n",
        "print(\"A sequ√™ncia transcrita √©\", rna)"
      ],
      "metadata": {
        "id": "OAHhNdaHtcEa"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "Por fim, podemos obter a sequ√™ncia de amino√°cidos a partir da tradu√ß√£o da sequ√™ncia:\n",
        "\n"
      ],
      "metadata": {
        "id": "CkHJcNaVth1q"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "protein = seq1.translate()\n",
        "print(\"A sequ√™ncia da prote√≠na √©\", protein)"
      ],
      "metadata": {
        "id": "pxHb5W1Qtisy"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## TRANSCRI√á√ÉO "
      ],
      "metadata": {
        "id": "amRe67d_xgIE"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "O processo de transcri√ß√£o permite a forma√ß√£o do RNA mensageiro com base na regi√£o codificante do DNA. Computacionalmente falando podemos analisar a transcri√ß√£o como um processo de modifica√ß√µes em strings."
      ],
      "metadata": {
        "id": "A6_xhxwWxlNP"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "from Bio.Seq import Seq\n",
        "\n",
        "dna = Seq(\"ATGGCCATTCGCAAGGGTGCCCGATAG\")\n",
        "print(\"DNA: \"+dna)\n",
        "# DNA: ATGGCCATTCGCAAGGGTGCCCGATAG\n",
        "\n",
        "rna = dna.transcribe()\n",
        "print(\"RNA: \"+rna)\n",
        "#RNA: AUGGCCAUUCGCAAGGGUGCCCGAUAG\n",
        "\n",
        "dna2 = rna.back_transcribe()\n",
        "print(\"DNA: \"+dna2)\n",
        "#DNA: ATGGCCATTCGCAAGGGTGCCCGATAG"
      ],
      "metadata": {
        "id": "dV3Qb8k9xoBf"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "# <center> Exemplo simples de an√°lise FASTA\n",
        "\n",
        "[exemplo](http://biopython.org/DIST/docs/tutorial/Tutorial.html)\n",
        "\n",
        "Se voc√™ abrir o arquivo FASTA de orqu√≠deas **ls_orchid.fasta** em seu editor de texto favorito, ver√° que o arquivo come√ßa assim:\n",
        "\n",
        "`>gi|2765658|emb|Z78533.1|CIZ78533 C.irapeanum 5.8S rRNA gene e ITS1 e ITS2 DNA\n",
        "CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGATGAGACCGTGGAATAAACGATCGAGTG AATCCGGAGGACCGGTGTACTCAGCTCACCGGGGGCATTGCTCCCGTGGTGACCCTGATTTGTTGTTGGG`\n",
        "...\n",
        "\n",
        "Acesse o [arquivo fasta](https://raw.githubusercontent.com/biopython/biopython/master/Doc/examples/ls_orchid.fasta)\n",
        "\n",
        "ARQUIVO FASTA:\n",
        "\n",
        "* Come√ßa sempre com ‚Äú >‚Äù (s√≠mbolo de maior que) seguido pela sequ√™ncia em uma ou mais linhas.\n",
        "* FASTA √© o formato preferido para armazenar genomas de refer√™ncia.\n",
        "* FASTA √© usado para armazenar todos os tipos de sequ√™ncias de DNA, RNA ou prote√≠nas \n",
        "\n",
        "Formato b√°sico\n",
        "Aqui est√° o formato b√°sico:\n",
        "\n",
        "\n",
        "    >sequence1_short_name with optional additional info after whitespace\n",
        "    ACATCACCCCATAAACAAATAGGTTTGGTCCTAGCCTTTCTATTAGCTCTTAGTAAGATTACACATGCAA\n",
        "    GCATCCCCGTTCCAGTGAGTTCACCCTCTAAATCACCACGATCAAAAGGAACAAGCATCAAGCACGCAGC\n",
        "    AATGCAGCTCAAAACGCTTAGCCTAGCCACACCCCCACGGGAAACAGCAGTGAT\n",
        "    >sequence2_short_name with optional additional info after whitespace\n",
        "    GCCCCAAACCCACTCCACCTTACTACCAGACAACCTTAGCCAAACCATTTACCCAAATAAAGTATAGGCG\n",
        "    ATAGAAATTGAAACCTGGCGCAATAGATATAGTACCGCAAGGGAAAGATGAAAAATTATAACCAAGCATA\n",
        "    ATATAG\n",
        "\n",
        "___\n",
        "\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "IqOOO1yu4rJ9"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "[O artigo FASTA original de Bill Pearson. Esta √© a ferramenta de software que tornou o formato popular](https://www.sciencedirect.com/science/article/abs/pii/007668799083007V)"
      ],
      "metadata": {
        "id": "8tKRakBnVrh8"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#Instalar o Biopython\n",
        "!pip3 install biopython"
      ],
      "metadata": {
        "id": "uX1GDdSw6iDW"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#importar biblioteca \n",
        "\n",
        "from Bio import SeqIO"
      ],
      "metadata": {
        "id": "xju3fPwi6WFE"
      },
      "execution_count": 7,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        " **Aqui est√° uma sequ√™ncia de refer√™ncia do mundo real (o genoma mitocondrial humano ) no formato FASTA:**"
      ],
      "metadata": {
        "id": "9M1XEAmWNuNt"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "https://colab.research.google.com/drive/1BwXFamKpKIhE5k2nFR8HpPvD4-00Z58s#scrollTo=uU3JCNKwMKgA\n",
        "\n",
        "https://biopython.org/wiki/SeqIO"
      ],
      "metadata": {
        "id": "vrTxvcWlOix2"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "aeoLgbAgMKf9"
      },
      "outputs": [],
      "source": [
        "import gzip\n",
        "import urllib.request\n",
        "url = 'ftp://ftp.ncbi.nlm.nih.gov/genomes/archive/old_genbank/Eukaryotes/vertebrates_mammals/Homo_sapiens/GRCh38/non-nuclear/assembled_chromosomes/FASTA/chrMT.fa.gz'\n",
        "response = urllib.request.urlopen(url)\n",
        "print(gzip.decompress(response.read()).decode('UTF8'))"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Este arquivo FASTA mostrado acima possui apenas uma sequ√™ncia. Como vimos no primeiro exemplo acima, tamb√©m √© poss√≠vel que um arquivo FASTA contenha v√°rias sequ√™ncias. Estes s√£o √†s vezes chamados de arquivos multi-FASTA. Quando voc√™ escreve c√≥digo para interpretar arquivos FASTA, √© uma boa ideia sempre permitir a possibilidade de que o arquivo FASTA possa conter v√°rias sequ√™ncias.\n",
        "\n",
        "Os arquivos FASTA geralmente s√£o armazenados com a `.fa` extens√£o do nome do arquivo, mas isso n√£o √© uma regra. `.fasta` √© outra extens√£o popular. Voc√™ tamb√©m pode ver `.fas`, `.fna`, `.mfa`(para multi-FASTA) e outros."
      ],
      "metadata": {
        "id": "qUywFGdWOR_u"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "---\n",
        "**Analisando o FASTA** ü§ì\n",
        "\n",
        "Aqui est√° uma fun√ß√£o simples para analisar um arquivo FASTA em um dicion√°rio Python. O dicion√°rio mapeia nomes curtos para cadeias de nucleot√≠deos correspondentes (com espa√ßos em branco removidos)."
      ],
      "metadata": {
        "id": "AFYGXkFF_APQ"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "def parse_fasta(fh):\n",
        "    fa = {}\n",
        "    current_short_name = None\n",
        "    # Part 1: compile list of lines per sequence\n",
        "    for ln in fh:\n",
        "        if ln[0] == '>':\n",
        "            # new name line; remember current sequence's short name\n",
        "            long_name = ln[1:].rstrip()\n",
        "            current_short_name = long_name.split()[0]\n",
        "            fa[current_short_name] = []\n",
        "        else:\n",
        "            # append nucleotides to current sequence\n",
        "            fa[current_short_name].append(ln.rstrip())\n",
        "    # Part 2: join lists into strings\n",
        "    for short_name, nuc_list in fa.items():\n",
        "        # join this sequence's lines into one long string\n",
        "        fa[short_name] = ''.join(nuc_list)\n",
        "    return fa"
      ],
      "metadata": {
        "id": "xzdHL97O_Mzi"
      },
      "execution_count": 13,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "from io import StringIO\n",
        "fasta_example = StringIO()\n",
        "parsed_fa = parse_fasta(fasta_example)\n",
        "parsed_fa"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "VAtw2-vhPy0P",
        "outputId": "330affcc-36e9-4f05-d871-a9034602270c"
      },
      "execution_count": 14,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "{}"
            ]
          },
          "metadata": {},
          "execution_count": 14
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Indexado FASTA**\n",
        "\n",
        "Digamos que voc√™ tenha um ou mais arquivos FASTA grandes (por exemplo, todo o genoma de refer√™ncia humano) e gostaria de acessar esses arquivos \"aleatoriamente\", espiando substrings aqui e ali sem nenhum padr√£o de acesso regular. \n",
        "\n",
        "Talvez voc√™ esteja imitando uma m√°quina de sequenciamento, lendo trechos de DNA aqui e ali.\n",
        "\n",
        "Voc√™ pode come√ßar usando a parse_fastafun√ß√£o definida acima para analisar os arquivos FASTA. Ent√£o, para acessar uma substring, fa√ßa o seguinte:"
      ],
      "metadata": {
        "id": "f0YgWno6RjB3"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "parsed_fa['sequence1'][100:130]"
      ],
      "metadata": {
        "id": "CaV3V3eIRcgm"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "#Agora Vamos explorar outro conjunto de Dados"
      ],
      "metadata": {
        "id": "KQPgTI5Ptxsm"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "Observe que o arquivo cont√©m 94 registros, cada um com uma linha come√ßando com ‚Äú >‚Äù (s√≠mbolo de maior que) seguido pela sequ√™ncia em uma ou mais linhas. \n",
        "\n",
        "\n",
        "Agora tente isso em Python:"
      ],
      "metadata": {
        "id": "s5IgopqONkpe"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#instalar\n",
        "!pip3 install pyfaidx"
      ],
      "metadata": {
        "id": "YZ_zQCsVWsb8"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#importar o arquivo FASTA\n",
        "from google.colab import files\n",
        "files.upload()"
      ],
      "metadata": {
        "id": "KeaK0vEB70W3"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "from pyfaidx import Fasta\n",
        "orquideas = Fasta('orquidea.fasta')"
      ],
      "metadata": {
        "id": "2lu0lSMeWVIy"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "orquideas.keys()"
      ],
      "metadata": {
        "id": "CgNFrmJXXGyC"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#explorar o index em uma sequ√™ncia espec√≠fica\n",
        "orquideas['gi|2765657|emb|Z78532.1|CCZ78532'][10:50]"
      ],
      "metadata": {
        "id": "_y_ZqeYOuG2J"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "orquideas['gi|2765657|emb|Z78532.1|CCZ78532'][10:50].seq"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 35
        },
        "id": "O9F78cQPwk24",
        "outputId": "19941593-aaff-4b24-aab4-346e056b15fa"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "'TTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACA'"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            }
          },
          "metadata": {},
          "execution_count": 8
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "orquideas['gi|2765657|emb|Z78532.1|CCZ78532'][10:50].name"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 35
        },
        "id": "0wSwH6ESwuGr",
        "outputId": "6788a0e7-94a7-47a6-d8b0-335e4e986e50"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "'gi|2765657|emb|Z78532.1|CCZ78532'"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            }
          },
          "metadata": {},
          "execution_count": 9
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "orquideas['gi|2765657|emb|Z78532.1|CCZ78532'][10:50].start"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "KSHrdMYswwFJ",
        "outputId": "ac7219c2-1080-4454-a06d-bf704c90a787"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "11"
            ]
          },
          "metadata": {},
          "execution_count": 10
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "orquideas['gi|2765657|emb|Z78532.1|CCZ78532'][10:50].end"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "m6M1Cm0owzvj",
        "outputId": "37eecafe-c956-483b-dcaf-b714c1575518"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "50"
            ]
          },
          "metadata": {},
          "execution_count": 11
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "orquideas['gi|2765657|emb|Z78532.1|CCZ78532'][10:50].fancy_name"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 35
        },
        "id": "RgORBV6Xw3yh",
        "outputId": "17598da6-4e5c-4091-a61b-6b03d62ced6d"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "'gi|2765657|emb|Z78532.1|CCZ78532:11-50'"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            }
          },
          "metadata": {},
          "execution_count": 12
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "len(orquideas['gi|2765657|emb|Z78532.1|CCZ78532'][10:50])"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "-gu7tFrWxCEv",
        "outputId": "a2c3f935-8140-4caa-dbd8-3d2fcf7b5a59"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "40"
            ]
          },
          "metadata": {},
          "execution_count": 13
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "https://github.com/mdshw5/pyfaidx"
      ],
      "metadata": {
        "id": "pYO9yMb-xGdF"
      }
    },
    {
      "cell_type": "markdown",
      "source": [
        "##Revisando"
      ],
      "metadata": {
        "id": "IjvNhspBHd1x"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "sequence = Seq('ATGGCAAAAGAGTGA')\n",
        "\n",
        "print('Sequence      (DNA):  ', str(sequence))\n",
        "print('Transcription (RNA):  ', str(sequence.transcribe()))\n",
        "print('Translation (Protein):', str(sequence.translate()))"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "rO9noSyTHgSE",
        "outputId": "703255bf-15e1-47c6-e074-a650fc3efd9d"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Sequence      (DNA):   ATGGCAAAAGAGTGA\n",
            "Transcription (RNA):   AUGGCAAAAGAGUGA\n",
            "Translation (Protein): MAKE*\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        ""
      ],
      "metadata": {
        "id": "wh3hDAC2HrFm"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}